File: KakaoBankApp/modules/aes-crypto/android/src/main/java/expo/modules/aescrypto/AesCryptoModule.kt
package expo.modules.aescrypto

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import android.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec

class AesCryptoException(message: String) : CodedException(message)

class AesCryptoModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AesCrypto")

    // Decrypts Base64 ciphertext with AES-128-ECB (PKCS5/PKCS7) using Android javax.crypto.
    Function("decryptAes128Ecb") { ciphertextBase64: String, secretKey: String ->
      decryptAes128Ecb(ciphertextBase64, secretKey)
    }
  }

  private fun decryptAes128Ecb(ciphertextBase64: String, secretKey: String): String {
    if (ciphertextBase64.isEmpty()) {
      throw AesCryptoException("Ciphertext is required")
    }

    val cipherBytes = try {
      Base64.decode(ciphertextBase64, Base64.DEFAULT)
    } catch (error: IllegalArgumentException) {
      throw AesCryptoException("Ciphertext must be valid Base64")
    }

    val keyBytes = makeAes128Key(secretKey)

    return try {
      val cipher = Cipher.getInstance("AES/ECB/PKCS5Padding")
      cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(keyBytes, "AES"))
      val plainBytes = cipher.doFinal(cipherBytes)
      String(plainBytes, Charsets.UTF_8)
    } catch (error: Exception) {
      throw AesCryptoException(error.message ?: "AES decrypt failed")
    }
  }

  private fun makeAes128Key(secretKey: String): ByteArray {
    val trimmed = secretKey.trim()
    if (trimmed.isEmpty()) {
      throw AesCryptoException("Secret key is required")
    }

    val isHex = trimmed.length == 32 && trimmed.matches(Regex("^[0-9a-fA-F]+$"))
    if (isHex) {
      val bytes = ByteArray(16)
      for (i in 0 until 16) {
        val index = i * 2
        bytes[i] = trimmed.substring(index, index + 2).toInt(16).toByte()
      }
      return bytes
    }

    val utf8Key = trimmed.toByteArray(Charsets.UTF_8)
    if (utf8Key.size != 16) {
      throw AesCryptoException(
        "AES-128 key must be exactly 16 bytes (got ${utf8Key.size})"
      )
    }
    return utf8Key
  }
}

File: KakaoBankApp/modules/aes-crypto/ios/AesCryptoModule.swift
import CommonCrypto
import ExpoModulesCore
import Foundation

internal final class AesCryptoException: Exception {
  private let message: String

  init(_ message: String) {
    self.message = message
    super.init()
  }

  override var reason: String {
    message
  }
}

public class AesCryptoModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AesCrypto")

    /// Decrypts Base64 ciphertext with AES-128-ECB (PKCS7) using iOS CommonCrypto.
    Function("decryptAes128Ecb") { (ciphertextBase64: String, secretKey: String) -> String in
      try Self.decrypt(ciphertextBase64: ciphertextBase64, secretKey: secretKey)
    }
  }

  private static func decrypt(ciphertextBase64: String, secretKey: String) throws -> String {
    guard !ciphertextBase64.isEmpty else {
      throw AesCryptoException("Ciphertext is required")
    }
    guard let cipherData = Data(base64Encoded: ciphertextBase64) else {
      throw AesCryptoException("Ciphertext must be valid Base64")
    }

    let keyData = try makeAes128Key(from: secretKey)
    let plainData = try crypt(data: cipherData, key: keyData)
    guard let plainText = String(data: plainData, encoding: .utf8) else {
      throw AesCryptoException("Decrypted payload is not valid UTF-8")
    }
    return plainText
  }

  private static func makeAes128Key(from secretKey: String) throws -> Data {
    let trimmed = secretKey.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
      throw AesCryptoException("Secret key is required")
    }

    let isHex = trimmed.count == 32 && trimmed.range(
      of: "^[0-9a-fA-F]+$",
      options: .regularExpression
    ) != nil

    if isHex {
      var bytes = [UInt8]()
      bytes.reserveCapacity(16)
      var index = trimmed.startIndex
      while index < trimmed.endIndex {
        let next = trimmed.index(index, offsetBy: 2)
        let byteString = trimmed[index..<next]
        guard let byte = UInt8(byteString, radix: 16) else {
          throw AesCryptoException("Hex AES key is invalid")
        }
        bytes.append(byte)
        index = next
      }
      guard bytes.count == kCCKeySizeAES128 else {
        throw AesCryptoException("Hex AES key must decode to exactly 16 bytes")
      }
      return Data(bytes)
    }

    guard let utf8Key = trimmed.data(using: .utf8) else {
      throw AesCryptoException("Secret key must be UTF-8")
    }
    guard utf8Key.count == kCCKeySizeAES128 else {
      throw AesCryptoException(
        "AES-128 key must be exactly 16 bytes (got \(utf8Key.count))"
      )
    }
    return utf8Key
  }

  private static func crypt(data: Data, key: Data) throws -> Data {
    let bufferSize = data.count + kCCBlockSizeAES128
    var outBytes = [UInt8](repeating: 0, count: bufferSize)
    var outLength: size_t = 0

    let status = key.withUnsafeBytes { keyBytes in
      data.withUnsafeBytes { dataBytes in
        CCCrypt(
          CCOperation(kCCDecrypt),
          CCAlgorithm(kCCAlgorithmAES),
          CCOptions(kCCOptionECBMode | kCCOptionPKCS7Padding),
          keyBytes.baseAddress,
          key.count,
          nil,
          dataBytes.baseAddress,
          data.count,
          &outBytes,
          bufferSize,
          &outLength
        )
      }
    }

    guard status == kCCSuccess else {
      throw AesCryptoException("AES decrypt failed with status \(status)")
    }

    return Data(bytes: outBytes, count: outLength)
  }
}
